libtenzir/builtins/operators/load_tcp.cpp

```diff
@@ -0,0 +1,981 @@
+  static constexpr auto has_terminal = false;
+  bool is_hidden = {};
+  node_actor node = {};
+
+  connection_manager_state() = default;
+  connection_manager_state(const connection_manager_state&) = delete;
+  connection_manager_state(connection_manager_state&&) = delete;
+  auto operator=(const connection_manager_state&)
+    -> connection_manager_state& = delete;
+  auto operator=(connection_manager_state&&)
+    -> connection_manager_state& = delete;
+
+  ~connection_manager_state() noexcept {
+    io_ctx->stop();
+    for (auto& io_worker : io_workers) {
+      io_worker.join();
+    }
+  }
+
+  auto start() -> caf::expected<void> {
+    auto tcp_metrics_schema = type{
+      "tenzir.metrics.tcp",
+      record_type{
+        {"handle", string_type{}},
+        {"reads", uint64_type{}},
+        {"writes", uint64_type{}},
+        {"bytes_read", uint64_type{}},
+        {"bytes_written", uint64_type{}},
+      },
+    };
+    self
+      ->request(metrics_receiver, caf::infinite, operator_id, tcp_metrics_id,
+                std::move(tcp_metrics_schema))
+      .then([]() {},
+            [this](const caf::error& err) {
+              diagnostic::error(err)
+                .note("failed to register TCP metrics schema")
+                .emit(diagnostics);
+            });
+    TENZIR_ASSERT(not io_ctx);
+    io_ctx = std::make_shared<boost::asio::io_context>();
+    io_workers.reserve(args.parallel.inner);
+    std::ranges::generate_n(
+      std::back_inserter(io_workers), args.parallel.inner, [this] {
+        return std::thread{[this] {
+          const auto guard = boost::asio::make_work_guard(*io_ctx);
+          io_ctx->run();
+        }};
+      });
+    socket.emplace(*io_ctx);
+    return args.connect ? connect() : listen();
+  }
+
+  auto connect() -> caf::expected<void> {
+    TENZIR_ASSERT(args.connect);
+    // TODO: Implement support for connect=true.
+    TENZIR_UNIMPLEMENTED();
+    return {};
+  }
+
+  auto listen() -> caf::expected<void> {
+    TENZIR_ASSERT(not args.connect);
+    TENZIR_ASSERT(not acceptor);
+    TRY(auto endpoint, resolve_endpoint(*io_ctx, args.endpoint));
+    acceptor.emplace(*io_ctx);
+    auto ec = boost::system::error_code{};
+    if (acceptor->open(endpoint.protocol(), ec)) {
+      return diagnostic::error("{}", ec.message())
+        .note("failed to open acceptor")
+        .primary(args.endpoint.source)
+        .to_error();
+    }
+    if (acceptor->set_option(boost::asio::socket_base::reuse_address{true},
+                             ec)) {
+      return diagnostic::error("{}", ec.message())
+        .note("failed to enable reuse address")
+        .primary(args.endpoint.source)
+        .to_error();
+    }
+    if (acceptor->bind(endpoint, ec)) {
+      return diagnostic::error("{}", ec.message())
+        .note("failed to bind to endpoint")
+        .primary(args.endpoint.source)
+        .to_error();
+    }
+    TRY(set_close_on_exec(acceptor->native_handle()));
```

Comment:
Not critical for this PR, but as discussed this can be moved up to reduce the race window.

Reply:
Good call, I left a comment.